sensing/autoware_pointcloud_preprocessor/include/autoware/pointcloud_preprocessor/outlier_filter/polar_voxel_outlier_filter_node.hpp

```diff
@@ -0,0 +1,194 @@
+
+  /** \brief Structure to represent a polar voxel index */
+  struct PolarVoxelIndex
+  {
+    int radius_idx;
+    int azimuth_idx;
+    int elevation_idx;
+
+    bool operator<(const PolarVoxelIndex & other) const
+    {
+      if (radius_idx != other.radius_idx) return radius_idx < other.radius_idx;
+      if (azimuth_idx != other.azimuth_idx) return azimuth_idx < other.azimuth_idx;
+      return elevation_idx < other.elevation_idx;
+    }
+  };
+
+  /** \brief Structure to hold both primary and secondary returns for each voxel */
+  struct VoxelPoints
+  {
+    std::vector<size_t> primary_returns;    // Points classified as primary returns
+    std::vector<size_t> secondary_returns;  // Points classified as secondary returns
+  };
+
+  /** \brief Filter using regular PointXYZ (compute polar coordinates)
+   * Uses simple occupancy threshold filtering - voxels with >= voxel_points_threshold points are
+   * kept. No return type classification or ratio-based filtering is applied.
+   */
+  void filter_point_xyz(
+    const PointCloud2ConstPtr & input, const IndicesPtr & indices, PointCloud2 & output);
+
+  /** \brief Filter using PointXYZIRCAEDT (use pre-computed polar coordinates)
+   * Uses advanced two-criteria filtering when return type classification is enabled:
+   * 1. Primary returns >= voxel_points_threshold
+   * 2. Primary-to-secondary ratio >= secondary_noise_threshold
+   * Both criteria must be satisfied for a voxel to be kept.
+   */
+  void filter_point_xyzircaedt(
+    const PointCloud2ConstPtr & input, const IndicesPtr & indices, PointCloud2 & output);
+
+  /** \brief Convert Cartesian coordinates to polar voxel index */
+  PolarVoxelIndex cartesian_to_polar_voxel(double x, double y, double z) const;
+
+  /** \brief Convert Cartesian point to polar coordinates */
+  static void cartesian_to_polar(
+    double x, double y, double z, double & radius, double & azimuth, double & elevation);
+
+  /** \brief Convert pre-computed polar coordinates to voxel index */
+  PolarVoxelIndex polar_to_polar_voxel(double radius, double azimuth, double elevation) const;
+
+  /** \brief Validate polar coordinates */
+  bool validate_polar_coordinates(double radius, double azimuth, double elevation) const;
+
+  /** \brief Check if return type is in primary returns list */
+  bool is_primary_return_type(uint8_t return_type) const;
+
+  /** \brief Check if return type is in secondary returns list */
+  bool is_secondary_return_type(uint8_t return_type) const;
+
+  /** \brief Classify point by return type for voxel mapping
+   * Classifies points into primary or secondary returns based on configured return type lists.
+   * Points not matching either list are ignored when return type classification is enabled.
+   */
+  void classify_point_by_return_type(
+    uint8_t return_type, size_t point_idx, PolarVoxelIndex voxel_idx,
+    std::map<PolarVoxelIndex, VoxelPoints> & voxel_point_map) const;
```

Comment:
Just from  the description, it is not apparent to me why the `voxel_point_map` is needed here, or what actually happens inside the functions. Please consider adding `@param` documentation lines.

Reply:
This got dissolved in the large-ish refactor into single vector allocation.